felix/daemon/daemon.go

```diff
@@ -267,11 +269,35 @@ configRetry:
 			continue configRetry
 		}
 
+		// Get all IP Pools to populate encapInfo.IPIPPools and encapInfo.VXLANPools
+		// and calculate encapInfo.UseIPIPEncap and encapInfo.UseVXLANEncap
+		ippoolKVList, err := backendClient.List(ctx, model.ResourceListOptions{Kind: apiv3.KindIPPool}, "")
+		if err != nil {
+			log.WithError(err).Error("Failed to list IP Pools")
+			time.Sleep(1 * time.Second)
+			continue configRetry
+		}
+		for _, kvp := range ippoolKVList.KVPairs {
+			pool := kvp.Value.(*apiv3.IPPool)
+			poolKey := pool.Spec.CIDR
+			if pool.Spec.IPIPMode != apiv3.IPIPModeNever {
+				encapInfo.IPIPPools[poolKey] = struct{}{}
+			}
+			if pool.Spec.VXLANMode != apiv3.VXLANModeNever {
+				encapInfo.VXLANPools[poolKey] = struct{}{}
+			}
+		}
+		encapInfo.UseIPIPEncap, encapInfo.UseVXLANEncap = calc.CalcIPIPVXLANEncaps(configParams.IpInIpEnabled, configParams.VXLANEnabled, encapInfo.IPIPPools, encapInfo.VXLANPools)
```

Comment:
The missing bit would be the restart callback, which we'd need to figure out. I think it's a bit messy to be attaching that to the encapInfo struct anyway, so would be nice to find a better way to pass that back up. Maybe making the RestartCallback a first-class argument to the CalculationGraph.

Reply:
This does seem more neat, thanks! The only note I'd say is that we would always iterate over the IP pools this way instead of keeping a map, right? That shouldn't be a problem, but just making sure...

Reply:
In fact, I'm still using maps the way I'm trying to go. Will push the changes as soon as I can.